fw/websocket.c

```diff
@@ -163,6 +163,11 @@ tfw_ws_msg_process(TfwConn *conn, struct sk_buff *skb)
 
 	T_DBG2("%s: conn=[%p], skb=[%p]\n", __func__, conn, skb);
 
+	if (!tfw_srv_conn_get_if_live(conn))
+		return TFW_BLOCK;
+	if (!tfw_srv_conn_get_if_live(conn->pair))
+		return TFW_BLOCK;
```

Comment:
I don't think that this is reliable. `tfw_ws_conn_drop()` calls `tfw_connection_put()`, which can just free `conn`, so if we have a race between the function and `conn` isn't protected, then we can just try to increment the refcaunter in a dead memory area.

The main question is who calls `tfw_ws_conn_drop()` - the server shutdown process (process context) or a socket error event (softirq with locked socket)? I guess the second one. The first one is easy to protect from - there just should not be active underlying sockets, when it frees the connections.

The second question is who exactly freed due to the race - `conn` or `conn->pair`? I think `conn->pair` since we're should be under socket lock for `conn`. If I'm right, then we should use the same mechanism as for forwarding responses to client connections - client connections can also die from under us.